File: cloudfunctions/plan/llm.js
'use strict'

function requireEnv(name) {
  const v = String(process.env[name] || '').trim()
  if (!v) {
    const err = new Error(`${name} is not set`)
    err.code = 'LLM_CONFIG'
    throw err
  }
  return v
}

async function chatComplete(messages, temperature = 0.6) {
  const apiKey = requireEnv('LLM_API_KEY')
  const baseUrl = requireEnv('LLM_BASE_URL')
  const model = requireEnv('LLM_MODEL')

  const res = await fetch(baseUrl, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${apiKey}`,
      'Content-Type': 'application/json',
      Accept: 'application/json',
    },
    body: JSON.stringify({ model, messages, temperature }),
  })

  let data
  try {
    data = await res.json()
  } catch {
    data = null
  }

  if (!res.ok) {
    const detail = data != null ? JSON.stringify(data) : await res.text()
    const err = new Error(`LLM HTTP ${res.status} for ${baseUrl}: ${detail}`)
    err.code = 'LLM_UPSTREAM'
    throw err
  }

  const choice0 = (data && data.choices && data.choices[0]) || {}
  let content = (((choice0.message || {}).content || '') + '').trim()
  if (!content) {
    content = (
      (((choice0.delta || {}).content || '') + '').trim() ||
      String(choice0.text || '').trim() ||
      String((data && data.output_text) || '').trim()
    )
  }
  if (!content) {
    const err = new Error(`LLM returned empty content: ${JSON.stringify(data)}`)
    err.code = 'LLM_UPSTREAM'
    throw err
  }
  return content
}

module.exports = { chatComplete }

File: backend/llm.py
from __future__ import annotations

import os
from typing import Any

import httpx


class LlmConfigMissing(RuntimeError):
    pass


class LlmUpstreamError(RuntimeError):
    pass


def _get_env(name: str) -> str:
    v = (os.getenv(name) or "").strip()
    if not v:
        raise LlmConfigMissing(f"{name} is not set")
    return v


def chat_complete(messages: list[dict[str, str]], *, temperature: float = 0.6) -> str:
    api_key = _get_env("LLM_API_KEY")
    base_url = _get_env("LLM_BASE_URL")
    model = _get_env("LLM_MODEL")

    payload: dict[str, Any] = {
        "model": model,
        "messages": messages,
        "temperature": temperature,
    }

    try:
        with httpx.Client(timeout=25.0) as client:
            res = client.post(
                base_url,
                headers={
                    "Authorization": f"Bearer {api_key}",
                    "Content-Type": "application/json",
                    "Accept": "application/json",
                },
                json=payload,
            )
            if res.status_code >= 400:
                detail = ""
                try:
                    detail = str(res.json())
                except Exception:
                    detail = res.text
                raise LlmUpstreamError(
                    f"LLM HTTP {res.status_code} for {base_url}: {detail}"
                )
            data = res.json()
    except Exception as e:
        raise LlmUpstreamError(f"Failed to call LLM: {e}") from e

    content = (
        (((data or {}).get("choices") or [{}])[0].get("message") or {}).get("content")
        or ""
    )
    if not content:
        # Try a few common fallbacks across OpenAI-compatible providers.
        choice0 = ((data or {}).get("choices") or [{}])[0] if isinstance((data or {}).get("choices"), list) else {}
        content = (
            (((choice0 or {}).get("delta") or {}).get("content") or "")
            or ((choice0 or {}).get("text") or "")
            or ((data or {}).get("output_text") or "")
        )

    if not content:
        raise LlmUpstreamError(f"LLM returned empty content: {data}")
    return content

